core/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java

```diff
@@ -128,6 +128,11 @@ boolean validate(MetaData metaData) {
                     throw new IllegalStateException("shard routing has an index [" + shardRouting.index() + "] that is different " +
                                                     "from the routing table");
                 }
+                if (shardRouting.active() &&
+                    indexMetaData.activeAllocationIds(shardRouting.id()).contains(shardRouting.allocationId().getId()) == false) {
+                        throw new IllegalStateException("active shard routing " + shardRouting + " has no corresponding entry in the " +
```

Comment:
this is never called from user-facing code. This method is only called by an assertion in AllocationService. I think we should change this method to use asserts instead of IllegalStateException, but that is another change, not in the scope of this PR.